<launch>
  <arg name="ocs" default="true" />
  <arg name="USE_SIMULATION" default="false" />
  <remap from="/tf" to="/ocs/tf" if="$(arg ocs)"/>
  <remap from="/joint_states" to="/ocs/joint_states" if="$(arg ocs)"/>
  <remap from="/robot_description" to="/ocs/robot_description" if="$(arg ocs)"/>
  <arg name="locomotion_cloud" default="/ocs/communication/locomotion_cloud" if="$(arg ocs)" />
  <arg name="locomotion_cloud" default="/locomotion/normal_filtered_points/output" unless="$(arg ocs)" />
  <arg name="flip_cloud" default="/ocs/communication/normal_flip_cloud" if="$(arg ocs)" />
  <arg name="flip_cloud" default="/locomotion/normal_flip/output" unless="$(arg ocs)" />

  <group ns="ocs">
    <node pkg="jsk_topic_tools" type="standalone_complexed_nodelet"
          name="manager" respawn="true" output="screen">
      <rosparam subst_value="true">
nodelets:
  - name: plane_estimation
    type: jsk_pcl/RegionGrowingMultiplePlaneSegmentation
    remappings:
      - from: ~input
        to: $(arg locomotion_cloud)
      - from: ~input_normal
        to: $(arg locomotion_cloud)
  - name: plane_concatenator
    type: jsk_pcl/PlaneConcatenator
    remappings:
      - from: ~input
        to: $(arg locomotion_cloud)
      - from: ~input/indices
        to: plane_estimation/output/inliers
      - from: ~input/polygons
        to: plane_estimation/output/polygons
      - from: ~input/coefficients
        to: plane_estimation/output/coefficients
  - name: obstacle_cloud
    type: pcl/ExtractIndices
    remappings:
      - from: ~input
        to: normal_flip/output
      - from: ~indices
        to: env_server/output/non_plane_indices
      </rosparam>
    </node>
    <group ns="plane_estimation">
      <rosparam>
        max_size: 50000
        min_size: 200
        angular_threshold: 0.01
        distance_threshold: 0.01
      </rosparam>
      <rosparam if="$(arg USE_SIMULATION)">
        cluster_tolerance: 0.05
      </rosparam>
    </group>
    <group ns="obstacle_cloud">
      <rosparam>
        negative: true
      </rosparam>
    </group>
    <node pkg="jsk_pcl_ros" type="environment_plane_modeling" name="env_server" output="screen">
      <remap from="~input" to="$(arg locomotion_cloud)"/>
      <remap from="~input/full_cloud" to="$(arg flip_cloud)" />
      <!-- <remap from="~input/full_cloud" to="$(arg locomotion_cloud)" /> -->
      <remap from="~input/polygons" to="plane_concatenator/output/polygons" />
      <remap from="~input/indices" to="plane_concatenator/output/indices" />
      <remap from="~input/coefficients" to="plane_concatenator/output/coefficients" />
      <remap from="~input/leg_bounding_box" to="/robot_legs_bbox" />
      <rosparam unless="$(arg USE_SIMULATION)">
        resolution: 0.01
        complete_footprint_region: false
        morphological_filter_size: 2
        # distance_threshold: 0.005
        distance_threshold: 0.025 # for jaxon
        magnify_distance: 0.05
        normal_threshold: 0.1
        erode_filter_size: 1
        footprint_frames: ["lleg_end_coords", "rleg_end_coords"]
      </rosparam>
      <rosparam if="$(arg USE_SIMULATION)">
        resolution: 0.01
        complete_footprint_region: false
        morphological_filter_size: 6
        # distance_threshold: 0.005
        distance_threshold: 0.025 # for jaxon
        magnify_distance: 0.05
        normal_threshold: 0.1
        erode_filter_size: 0
        footprint_frames: ["lleg_end_coords", "rleg_end_coords"]
      </rosparam>
    </node>
  </group>
</launch>
